ServerCore/Areas/Identity/UserAuthorizationPolicy/AuthorizationHelper.cs

```diff
@@ -0,0 +1,122 @@
+using Microsoft.AspNetCore.Mvc.Filters;
+using ServerCore.DataModel;
+using ServerCore.Helpers;
+
+namespace ServerCore.Areas.Identity
+{
+    public class AuthorizationHelper
+    {
+        public static Event GetEventFromContext(AuthorizationHandlerContext context)
+        {
+            if (context.Resource is AuthorizationFilterContext filterContext)
+            {
+                string eventIdAsString = filterContext.RouteData.Values["eventId"] as string;
+
+                if (Int32.TryParse(eventIdAsString, out int eventId))
+                {
+                    PuzzleServerContext puzzleServerContext = (PuzzleServerContext)filterContext.HttpContext.RequestServices.GetService(typeof(PuzzleServerContext));
+                    return puzzleServerContext.Events.Where(e => e.ID == eventId).FirstOrDefault();
+                }
+            }
+
+            return null;
+        }
+
+        public static Puzzle GetPuzzleFromContext(AuthorizationHandlerContext context)
+        {
+            if (context.Resource is AuthorizationFilterContext filterContext)
+            {
+                string puzzleIdAsString = filterContext.RouteData.Values["puzzleId"] as string;
+
+                if (Int32.TryParse(puzzleIdAsString, out int puzzleId))
+                {
+                    PuzzleServerContext puzzleServerContext = (PuzzleServerContext)filterContext.HttpContext.RequestServices.GetService(typeof(PuzzleServerContext));
+                    return puzzleServerContext.Puzzles.Where(e => e.ID == puzzleId).FirstOrDefault();
+                }
+            }
+
+            return null;
+        }
+
+        public static Team GetTeamFromContext(AuthorizationHandlerContext context)
+        {
+            if (context.Resource is AuthorizationFilterContext filterContext)
+            {
+                string teamIdAsString = filterContext.RouteData.Values["teamId"] as string;
+
+                if (Int32.TryParse(teamIdAsString, out int teamId))
+                {
+                    PuzzleServerContext puzzleServerContext = (PuzzleServerContext)filterContext.HttpContext.RequestServices.GetService(typeof(PuzzleServerContext));
+                    return puzzleServerContext.Teams.Where(e => e.ID == teamId).FirstOrDefault();
+                }
+            }
+
+            return null;
+        }
+
+        public static Task IsEventAdminCheck(AuthorizationHandlerContext authContext, PuzzleServerContext dbContext, UserManager<IdentityUser> userManager, IAuthorizationRequirement requirement)
+        {
+            PuzzleUser puzzleUser = PuzzleUser.GetPuzzleUserForCurrentUser(dbContext, authContext.User, userManager);
+
+            Event thisEvent = AuthorizationHelper.GetEventFromContext(authContext);
+
+            if (thisEvent != null && puzzleUser.IsAdminForEvent(dbContext, thisEvent).Result)
+            {
+                authContext.Succeed(requirement);
+            }
+
+            return Task.CompletedTask;
```

Comment:
If you're not awaiting anything, you could just make the method void. That said, I think it would make more sense to make GetPuzzleUserForCurrentUser, GetEventFromContext and IsAdminForEvent async, in which case, you can make this method async and still not need to return Task.CompletedTask. Same comment for the other checks.

Reply:
The whole set of checks here was set up this way because that's what happened when I used extract method. Got them all cleaned up.